packages/codecs-core/src/codec.ts

```diff
@@ -4,36 +4,63 @@
 export type Offset = number;
 
 /**
- * The shared attributes between codecs, encoders and decoders.
- */
-export type CodecData = {
-    /** A description for the codec. */
-    description: string;
-    /** The fixed size of the encoded value in bytes, or `null` if it is variable. */
-    fixedSize: number | null;
-    /** The maximum size an encoded value can be in bytes, or `null` if it is variable. */
-    maxSize: number | null;
+ * An object that can encode a value to a `Uint8Array`.
+ */
+export type Encoder<T> = EncoderSize<T> & {
+    /** Encode the provided value and return the encoded bytes directly. */
+    readonly encode: (value: T) => Uint8Array;
+    /**
+     * Writes the encoded value into the provided byte array at the given offset.
+     * Returns the offset of the next byte after the encoded value.
+     */
+    readonly write: (value: T, bytes: Uint8Array, offset: Offset) => Offset;
 };
 
 /**
- * An object that can encode a value to a `Uint8Array`.
+ * Describes the size of an Encoder, either fixed or variable.
  */
-export type Encoder<T> = CodecData & {
-    /** The function that encodes a value into bytes. */
-    encode: (value: T) => Uint8Array;
-};
+export type EncoderSize<T> =
+    | {
+          /** The fixed size of the encoded value in bytes, if applicable. */
+          readonly fixedSize: number;
+      }
+    | {
+          /** Otherwise, a null fixedSize indicates it's a variable size encoder. */
+          readonly fixedSize: null;
+          /** The maximum size an encoded value can be in bytes, if applicable. */
+          readonly maxSize?: number;
+          /** The total size of the encoded value in bytes. */
+          readonly variableSize: (value: T) => number;
```

Comment:
This might confuse people because ‘variable’ is both a noun and a verb. `getSizeForValue()`?

Reply:
I'm okay with `getSizeForValue`, I'll do the change in a separate PR.

Reply:
See https://github.com/solana-labs/solana-web3.js/pull/1901